src/hooks/useOrds.ts

```diff
@@ -598,6 +611,157 @@ export const useOrds = () => {
+      setIsProcessing(true);
+
+      const isAuthenticated = await verifyPassword(password);
+      if (!isAuthenticated) {
+        return { error: 'invalid-password' };
+      }
+      const keys = await retrieveKeys(password);
+
+      if (!keys.walletWif || !keys.ordWif) return { error: 'no-keys' };
+      const fundingAndChangeAddress = bsvAddress;
+
+      const fundingUtxos = await getUtxos(fundingAndChangeAddress);
+
+      if (!fundingUtxos.length) {
+        throw new Error('Could not retrieve funding UTXOs');
+      }
+
+      const payPk = PrivateKey.from_wif(keys.walletWif);
+      const res = await getUtxoByOutpoint(outpoint);
+      const { vout, txid } = res;
+      const price = Number(res.data?.list?.price);
+      const payout = res.data?.list?.payout;
+
+      const { script } = await getMarketData(outpoint);
+
+      if (!txid || !price || !payout || !script) throw Error('Missing information!');
+
+      const purchaseTx = new Transaction(1, 0);
+
+      const neededAmount = fundingUtxos.reduce((a, r) => a + r.satoshis, 0) + P2PKH_INPUT_SIZE * fundingUtxos.length;
+      if (price * (1 + marketplaceRate) >= neededAmount) {
+        return { error: 'insufficient-funds' };
+      }
+      const listingInput = new TxIn(Buffer.from(txid, 'hex'), vout, Script.from_asm_string(''));
+      purchaseTx.add_input(listingInput);
+
+      // output 0
+      const buyerOutput = new TxOut(BigInt(1), P2PKHAddress.from_string(ordAddress).get_locking_script());
+      purchaseTx.add_output(buyerOutput);
+
+      // output 1
+      const payOutput = TxOut.from_hex(Buffer.from(payout, 'base64').toString('hex'));
+      purchaseTx.add_output(payOutput);
+
+      // output 2 - change
+      const dummyChangeOutput = new TxOut(
+        BigInt(0),
+        P2PKHAddress.from_string(fundingAndChangeAddress).get_locking_script(),
+      );
+      purchaseTx.add_output(dummyChangeOutput);
+
+      // output 3 - marketFee
+      const dummyMarketFeeOutput = new TxOut(
+        BigInt(0),
+        P2PKHAddress.from_string(marketplaceAddress).get_locking_script(),
+      );
+      purchaseTx.add_output(dummyMarketFeeOutput);
+
+      let preimage = purchaseTx.sighash_preimage(
+        SigHash.InputOutput,
+        0,
+        Script.from_bytes(Buffer.from(script, 'base64')),
+        BigInt(1), //TODO: use amount from listing
+      );
+
+      listingInput.set_unlocking_script(
+        Script.from_asm_string(
+          `${purchaseTx.get_output(0)!.to_hex()} ${purchaseTx.get_output(2)!.to_hex()}${purchaseTx
+            .get_output(3)!
+            .to_hex()} ${Buffer.from(preimage).toString('hex')} OP_0`,
+        ),
+      );
+      purchaseTx.set_input(0, listingInput);
+
+      let marketFee = Math.ceil(price * marketplaceRate);
+      let fee = calculateFee(1, purchaseTx);
+      let satsNeeded = fee + price + marketFee;
+      const inputs = getInputs(fundingUtxos, satsNeeded, false);
```

Comment:
This will consume all of a user's funding Utxos. Should probably filter it down so you are only adding enough to pay for the transaction

Reply:
I was wrong. I misread how getInputs works. Still, input 0 will be MUCH larger than P2PKH_INPUT_SIZE